From a5d055fb05aec4b0b4ab75ebae04b0ef0064ff14 Mon Sep 17 00:00:00 2001
From: Mengkejiergeli Ba <mengkejiergeli.ba@intel.com>
Date: Wed, 2 Jun 2021 14:47:55 +0800
Subject: [PATCH 2/6] msdkav1enc: add 10bit p010

---
 sys/msdk/gstmsdkav1enc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sys/msdk/gstmsdkav1enc.c b/sys/msdk/gstmsdkav1enc.c
index 0d82914c7e..c4c5337888 100644
--- a/sys/msdk/gstmsdkav1enc.c
+++ b/sys/msdk/gstmsdkav1enc.c
@@ -40,7 +40,7 @@
 GST_DEBUG_CATEGORY_EXTERN (gst_msdkav1enc_debug);
 #define GST_CAT_DEFAULT gst_msdkav1enc_debug
 
-#define RAW_FORMATS "NV12, I420, YV12"
+#define RAW_FORMATS "NV12, I420, YV12, P010_10LE"
 #define PROFILES    "main"
 
 #define COMMON_FORMAT "{ " RAW_FORMATS " }"
@@ -121,6 +121,7 @@ gst_msdkav1enc_configure (GstMsdkEnc * encoder)
 
   switch (encoder->param.mfx.FrameInfo.FourCC) {
     case MFX_FOURCC_NV12:
+    case MFX_FOURCC_P010:
       encoder->param.mfx.CodecProfile = MFX_PROFILE_AV1_MAIN;
       break;
 
-- 
2.17.1

