From d563eee1e81de9bd54d4d56e78bc11900d92ec76 Mon Sep 17 00:00:00 2001
From: Mengkejiergeli Ba <mengkejiergeli.ba@intel.com>
Date: Tue, 29 Jun 2021 17:23:44 +0800
Subject: [PATCH 4/6] msdkav1enc: fix qp range

---
 sys/msdk/gstmsdkav1enc.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/sys/msdk/gstmsdkav1enc.c b/sys/msdk/gstmsdkav1enc.c
index c31a395006..534e57c97b 100644
--- a/sys/msdk/gstmsdkav1enc.c
+++ b/sys/msdk/gstmsdkav1enc.c
@@ -276,20 +276,22 @@ gst_msdkav1enc_class_init (GstMsdkAV1EncClass * klass)
   encoder_class->set_format = gst_msdkav1enc_set_format;
   encoder_class->configure = gst_msdkav1enc_configure;
   encoder_class->set_src_caps = gst_msdkav1enc_set_src_caps;
+  encoder_class->qp_max = 255;
+  encoder_class->qp_min = 0;
 
   gst_msdkenc_install_common_properties (encoder_class);
 
   g_object_class_install_property (gobject_class, PROP_TILE_ROW,
-      g_param_spec_uint ("num-tile-rows", "number of rows for tiled encoding",
+      g_param_spec_uint ("num-tile-rows",
           "number of rows for tiled encoding",
-          1, 8192, PROP_TILE_ROW_DEFAULT,
-          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+          "number of rows for tiled encoding", 1, 8192,
+          PROP_TILE_ROW_DEFAULT, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
   g_object_class_install_property (gobject_class, PROP_TILE_COL,
-      g_param_spec_uint ("num-tile-cols", "number of columns for tiled encoding",
+      g_param_spec_uint ("num-tile-cols",
           "number of columns for tiled encoding",
-          1, 8192, PROP_TILE_COL_DEFAULT,
-          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+          "number of columns for tiled encoding", 1, 8192,
+          PROP_TILE_COL_DEFAULT, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
   gst_element_class_set_static_metadata (element_class,
       "Intel MSDK AV1 encoder",
-- 
2.17.1

