From 4a0f8b9a7e431ebbdf56aa1c48f16ad2ff17b1d5 Mon Sep 17 00:00:00 2001
From: He Junyan <junyan.he@intel.com>
Date: Mon, 5 Jun 2023 15:42:11 +0800
Subject: [PATCH 111/127] gir: Update the gir for GstVa-1.0.gir

---
 girs/GstVa-1.0.gir | 32 +++++++++++++++++++++++++++++---
 1 file changed, 29 insertions(+), 3 deletions(-)

diff --git a/girs/GstVa-1.0.gir b/girs/GstVa-1.0.gir
index 910107ef28..b6f63a7215 100644
--- a/girs/GstVa-1.0.gir
+++ b/girs/GstVa-1.0.gir
@@ -672,9 +672,9 @@ otherwise.</doc>
         </parameters>
       </function>
       <function name="set_format" c:identifier="gst_va_dmabuf_allocator_set_format" version="1.22">
-        <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">Sets the configuration defined by @info and @usage_hint for
-@allocator, and it tries the configuration, if @allocator has not
-allocated memories yet.
+        <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">Sets the configuration defined by @info, @usage_hint and @modifier
+for @allocator, and it tries the configuration, if @allocator has
+not allocated memories yet.
 
 If @allocator has memory allocated already, and frame size and
 format in @info are the same as currently configured in @allocator,
@@ -698,6 +698,10 @@ configuration is not valid or not updated.</doc>
             <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">VA usage hint</doc>
             <type name="guint" c:type="guint"/>
           </parameter>
+          <parameter name="modifier" transfer-ownership="none">
+            <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">the underlying modifier</doc>
+            <type name="guint64" c:type="guint64"/>
+          </parameter>
         </parameters>
       </function>
       <function name="setup_buffer" c:identifier="gst_va_dmabuf_allocator_setup_buffer" version="1.22">
@@ -980,6 +984,28 @@ to it.
         </parameter>
       </parameters>
     </function>
+    <function name="va_dmabuf_get_modifier_for_format" c:identifier="gst_va_dmabuf_get_modifier_for_format" version="1.24">
+      <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">Get the underlying modifier for specified @format and @usage_hint.</doc>
+      <source-position filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.h"/>
+      <return-value transfer-ownership="none">
+        <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">the underlying modifier.</doc>
+        <type name="guint64" c:type="guint64"/>
+      </return-value>
+      <parameters>
+        <parameter name="display" transfer-ownership="none">
+          <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">a #GstVaDisplay</doc>
+          <type name="VaDisplay" c:type="GstVaDisplay*"/>
+        </parameter>
+        <parameter name="format" transfer-ownership="none">
+          <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">a #GstVideoFormat</doc>
+          <type name="GstVideo.VideoFormat" c:type="GstVideoFormat"/>
+        </parameter>
+        <parameter name="usage_hint" transfer-ownership="none">
+          <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">VA usage hint</doc>
+          <type name="guint" c:type="guint"/>
+        </parameter>
+      </parameters>
+    </function>
     <function name="va_dmabuf_memories_setup" c:identifier="gst_va_dmabuf_memories_setup" version="1.22" introspectable="0">
       <doc xml:space="preserve" filename="../subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c">It imports the array of @mem, representing a single frame, into a
 VASurfaceID and it's attached into every @mem.</doc>
-- 
2.25.1

