From 2fa8ace4c4f6b557390be1c6ca680cd96ebab060 Mon Sep 17 00:00:00 2001
From: Mengkejiergeli Ba <mengkejiergeli.ba@intel.com>
Date: Fri, 2 Jun 2023 17:30:31 +0800
Subject: [PATCH 137/138] msdk: Add a help func to fix the map

---
 .../gst-plugins-bad/sys/msdk/gstmsdkcaps.c    | 29 +++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/subprojects/gst-plugins-bad/sys/msdk/gstmsdkcaps.c b/subprojects/gst-plugins-bad/sys/msdk/gstmsdkcaps.c
index 5b036bf849..1ad4a00591 100644
--- a/subprojects/gst-plugins-bad/sys/msdk/gstmsdkcaps.c
+++ b/subprojects/gst-plugins-bad/sys/msdk/gstmsdkcaps.c
@@ -224,6 +224,34 @@ _get_media_type (guint codec)
 }
 
 #ifndef _WIN32
+/* In RGB case, we need to fix the map */
+static void
+_fix_map (GstMsdkContext * context)
+{
+  GstVaDisplay *display = NULL;
+  VAImageFormat *va_formats;
+  VADisplay dpy;
+  VAStatus status;
+  int max, num = 0;
+
+  display = (GstVaDisplay *) gst_msdk_context_get_va_display (context);
+  dpy = gst_va_display_get_va_dpy (display);
+
+  max = vaMaxNumImageFormats (dpy);
+  if (max == 0)
+    return;
+
+  va_formats = g_new (VAImageFormat, max);
+  status = vaQueryImageFormats (dpy, va_formats, &num);
+  gst_va_video_format_fix_map (va_formats, num);
+
+  if (status != VA_STATUS_SUCCESS) {
+    GST_WARNING ("vaQueryImageFormats: %s", vaErrorStr (status));
+    g_free (va_formats);
+  }
+  return;
+}
+
 static gboolean
 _dma_fmt_to_dma_drm_fmts (GstMsdkContext * context,
     GstMsdkContextJobType job_type,
@@ -252,6 +280,7 @@ _dma_fmt_to_dma_drm_fmts (GstMsdkContext * context,
   for (gint m = 0; m < gst_value_list_get_size (&mods); m++) {
     const GValue *gmod = gst_value_list_get_value (&mods, m);
     modifier = g_value_get_uint64 (gmod);
+    _fix_map (context);
     drm_fourcc = gst_va_drm_fourcc_from_video_format (fmt);
 
     if (modifier == DRM_FORMAT_MOD_INVALID || drm_fourcc == DRM_FORMAT_INVALID)
-- 
2.34.1

